File: mailblastr-python/mailblastr/audiences.py
"""Audiences resource (incl. Google Sheet contact import)."""

from typing import TypedDict

from . import http_client
from .http_client import paginate, path_escape as _e


class CreateParams(TypedDict):
    name: str


class ImportSheetParams(TypedDict, total=False):
    url: str  # required: link-shared Google Sheet URL
    segment_name: str


class Audiences:
    """``mailblastr.Audiences`` — the /audiences endpoints."""

    CreateParams = CreateParams
    ImportSheetParams = ImportSheetParams

    @classmethod
    def create(cls, params):
        """Create an audience. POST /audiences"""
        return http_client.request("POST", "/audiences", params)

    @classmethod
    def get(cls, audience_id):
        """Retrieve an audience. GET /audiences/:id"""
        return http_client.request("GET", f"/audiences/{_e(audience_id)}")

    @classmethod
    def list(cls, params=None):
        """List audiences. GET /audiences"""
        return http_client.request("GET", f"/audiences{paginate(params)}")

    @classmethod
    def update(cls, audience_id, params):
        """Rename an audience. PATCH /audiences/:id"""
        return http_client.request("PATCH", f"/audiences/{_e(audience_id)}", params)

    @classmethod
    def import_sheet(cls, audience_id, params):
        """Import contacts from a link-shared Google Sheet. Header columns
        become contact properties (usable as ``{{merge_tags}}``); rows land in
        a fresh segment. POST /audiences/:id/contacts/import-sheet"""
        return http_client.request(
            "POST", f"/audiences/{_e(audience_id)}/contacts/import-sheet", params
        )

    @classmethod
    def remove(cls, audience_id):
        """Delete an audience. DELETE /audiences/:id"""
        return http_client.request("DELETE", f"/audiences/{_e(audience_id)}")

File: mailblastr-java/src/main/java/com/mailblastr/resources/Audiences.java
package com.mailblastr.resources;

import com.mailblastr.ListParams;
import com.mailblastr.MailblastrResponse;
import com.mailblastr.http.ApiClient;

import java.util.LinkedHashMap;
import java.util.Map;

/** Audiences — {@code mailblastr.audiences()}. */
public final class Audiences extends Resource {
    public Audiences(ApiClient api) { super(api); }

    /** {@code POST /audiences} */
    public MailblastrResponse create(String name) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("name", name);
        return api.request("POST", "/audiences", body);
    }

    /** {@code GET /audiences/:id} */
    public MailblastrResponse get(String id) {
        return api.request("GET", "/audiences/" + enc(id));
    }

    /** {@code GET /audiences} */
    public MailblastrResponse list() { return list(null); }

    public MailblastrResponse list(ListParams params) {
        return api.request("GET", "/audiences" + paginate(params));
    }

    /**
     * Import contacts from a link-shared Google Sheet — header columns become
     * contact properties; rows land in a fresh segment.
     * {@code POST /audiences/:id/contacts/import-sheet}
     */
    public MailblastrResponse importSheet(String audienceId, String url) {
        return importSheet(audienceId, url, null);
    }

    public MailblastrResponse importSheet(String audienceId, String url, String segmentName) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("url", url);
        if (segmentName != null) body.put("segment_name", segmentName);
        return api.request("POST", "/audiences/" + enc(audienceId) + "/contacts/import-sheet", body);
    }

    /** Rename an audience. {@code PATCH /audiences/:id} */
    public MailblastrResponse update(String id, String name) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("name", name);
        return api.request("PATCH", "/audiences/" + enc(id), body);
    }

    /** {@code DELETE /audiences/:id} */
    public MailblastrResponse remove(String id) {
        return api.request("DELETE", "/audiences/" + enc(id));
    }
}
